python/lsst/ts/tcpip/one_client_read_loop_server.py

```diff
@@ -0,0 +1,143 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import abc
+import asyncio
+import logging
+import typing
+
+from .base_client_or_server import ConnectCallbackType
+from .one_client_server import OneClientServer
+
+__all__ = ["OneClientReadLoopServer"]
+
+
+class OneClientReadLoopServer(OneClientServer):
+    """A OneClientServer that reads and processes data in a loop.
+
+    Parameters
+    ----------
+    host : `str` or `None`
+        IP address for this server; typically `LOCALHOST_IPV4` for IPV4
+        or `LOCALHOST_IPV6` for IPV6. If `None` then bind to all network
+        interfaces (e.g. listen on an IPv4 socket and an IPv6 socket).
+        None can cause trouble with port=0;
+        see ``port`` in the Attributes section for more information.
+    port : `int`
+        IP port for this server. If 0 then randomly pick an available port
+        (or ports, if listening on multiple sockets).
+        0 is strongly recommended for unit tests.
+    log : `logging.Logger`
+        Logger.
+    connect_callback : callable or `None`, optional
+        Asynchronous or (deprecated) synchronous function to call when
+        when a client connects or disconnects.
+        If the other end (client) closes the connection, it may take
+        ``monitor_connection_interval`` seconds or longer to notice.
+        The function receives one argument: this `OneClientServer`.
+    name : `str`, optional
+        Name used for log messages, e.g. "Commands" or "Telemetry".
+    **kwargs : `dict` [`str`, `typing.Any`]
+        Additional keyword arguments for `asyncio.start_server`,
+        beyond host and port.
+
+    Attributes
+    ----------
+    read_loop_task : `asyncio.Future`
+        A task that reads messages from the reader in a loop.
+    """
+
+    def __init__(
+        self,
+        host: str | None,
+        port: int | None,
+        log: logging.Logger,
+        connect_callback: ConnectCallbackType | None = None,
+        name: str = "",
+        **kwargs: typing.Any,
+    ) -> None:
+        self.read_loop_task: asyncio.Future = asyncio.Future()
+
+        super().__init__(
+            host=host,
+            port=port,
+            log=log,
+            connect_callback=connect_callback,
+            monitor_connection_interval=0,
+            name=name,
+            **kwargs,
+        )
+
+    async def call_connect_callback(self) -> None:
+        """A client has connected or disconnected."""
+        await super().call_connect_callback()
+        if self.connected:
+            self.log.info("Client connected.")
+            self.read_loop_task.cancel()
+            self.read_loop_task = asyncio.create_task(self.read_loop())
```

Comment:
I am wondering if the user's connect callback is likely to care if the read loop is running. Some thoughts (all a bit unlikely):
* If the connect callback wants to check that the read loop successfully started/is still running, then this order is reversed -- the user's callback should be called last.
* If the connect callback needs to do something to *prepare* for the read loop, then this order is right.

I have no idea how likely either of these is. Do you?

Reply:
That one mention of "messages" slipped my attention and has been replaced with "data". I made sure that that was the last one.

Reply:
I also don't know how likely any of those connect callback scenarios is. I merely added the override so unit tests can inject their own callback which in some cases helps to simplify the unit test code. I didn't consider the possibility that implementing classes would want to implement their own callback method.

I propose we leave it as is and worry about this later if this ever turns out to be an issue. If you disagree then I can add empty pre-connect-callback and post-connect-callback methods that the implementing classes can override if they need to.

Reply:
I agree.